### ISSUE TITLE
deadlock in maintenance when applying additional constraints

### ISSUE THREAD

**Issue by lucasmo** [keithf4/pg_partman#91]
Let's say we have a table hierarchy:

```
table_current
table_previous
...
table_oldest
```

current, previous, and oldest are actually partition numbers but are simplified for the sake of this example.

We call run_maintenance, which performs the following steps:

1. A new table is created, table_new. This becomes the front of the partition set. This locks on this table do not matter because it's for future use.
1. Constraints are applied to table_previous, which acquires an AccessExclusiveLock on table_previous via ALTER TABLE ADD CONSTRAINT.
1. table_oldest is disinherited, which requires an AccessExclusiveLock on table_oldest via ALTER TABLE NO INHERIT.

But what if, between step 2 and step 3, another transaction starts that selects from our partition set? This could be any old data query.

1. An AccessShareLock is attempted on all partition members. An AccessShareLock is acquired for table_oldest, and the process sits waiting for an AccessShareLock on table_previous, which is exclusively locked.
1. A deadlock occurs, with maintenance trying to exclusively lock table_oldest to drop it and the data query trying to share lock table_previous.

I'm not sure of how to fix this aside from simply not using constraints, or perhaps by only applying constraints in a separate transaction via a separate maintenance function.

**Comment by lucasmo** [comment]
Example:
```
ERROR:  deadlock detected
CONTEXT: SQL statement "ALTER TABLE public.table_p5207000000 NO INHERIT public.table"
PL/pgSQL function partman.drop_partition_id(text,bigint,boolean,boolean,text) line 137 at EXECUTE statement
PL/pgSQL function partman.run_maintenance(text,boolean,boolean,boolean) line 321 at assignment
DETAIL: Process 20834 waits for AccessExclusiveLock on relation 50889059 of database 170275; blocked by process 21056.
Process 21056 waits for AccessShareLock on relation 53563399 of database 170275; blocked by process 20834.
```
```
# select relname from pg_class where oid = 50889059;
table_p5207000000
```
```
# select relname from pg_class where oid = 53563399;
table_p5505000000
```

**Comment by keithf4** [comment]
Hmm... yeah not sure how to fix it other than what you suggested. Perhaps a parameter to run_maintenace() to tell it whether or not to apply additional constraints during that run?

**Comment by keithf4** [comment]
Just thinking out loud for an opinion... what if the retention management was moved first in run_maintenance before new child tables are created?

Also, in the next update, I've actually moved the constraint creation out of the `run_maintenance()` function and into the `create_partition_*()` functions because of a bug I found where if constraints need to be applied on multiple old tables, only one of them was getting applied. Wondering if that may lead to additional race conditions, but can't find another way around it at the moment.

**Comment by broseed** [comment]
Any update on this issue?  I experience deadlock often when run_maintenance is called on very busy partitions (that may have multiple transactions/second). Any advice appreciated

**Comment by keithf4** [comment]
Unfortunately this is not an easy fix with normal plpgsql since all function calls are done in a single transaction. I'm working on a C version of the maintenance function that will hopefully be much more resilient to concurrency issues. I've got a lot to learn there, though, so it's taking time.

If you continue to have issues, my only suggestion at this point is to exclude the very busy partition sets from general `run_maintenance` calls (by setting the `use_run_maintenance` flag in the `part_config` table) and schedule those individually by passing that parent table as an argument to `run_maintenance` in its own call. That will ensure that it completes as quickly as possible and other partition maintenance will not hold it up.

Sorry for the problems, but people are using this extension in much larger situations than I ever imagined. It's taking time for me to learn the proper ways to mange things like this in high traffic situations. Appreciate your patience. :)

**Comment by broseed** [comment]
Understood - and thanks for this extension which has been great so far!

Would the explicit call to run_maintenance with the parent table name on a
high-traffic table not result in deadlocks?

**Comment by keithf4** [comment]
It may. But it should lessen the time of the overall maintenance run since it's only running for that one instance, which should lessen the chance of a deadlock.. If the run-time is still a problem, you can try passing the `p_analyze` parameter as false. But, as the documentation states, it's recommended you schedule an analyze somehow outside of the maintenance run then so that constraint exclusion works properly.

**Comment by keithf4** [comment]
I'd left this issue open because I wanted to not forget about it and maybe look into possibly making more maintenance operations in C. But it looks like PG11 may have a solution to this and I've opened up a new issue for it:  https://github.com/pgpartman/pg_partman/issues/219

**Comment by keithf4** [comment]
So this issue likely isn't going to be possible to fix on any version of PG except 11+ and using stored procedures. I've started some initial work with this by creating a `run_maintenance_proc()` procedure that automatically commits after each partition set's maintenance has done. This should help reduce contention quite a bit by reducing the transaction time of the maintenance runs.

However, the race condition originally mentioned in this issue isn't going to be solved until the maintenance process is broken down even further into smaller, commitable transactions. I'm hoping I have the time to work on this in the future, but don't have an ETA for it at this time.
